Add MISRA.md (#35)
FreeRTOS/coreMQTT-Agent: MISRA.md

# MISRA Compliance

The coreMQTT Agent library files conform to the [MISRA C:2012](https://www.misra.org.uk/MISRAHome/MISRAC2012/tabid/196/Default.aspx)
guidelines, with some noted exceptions. Compliance is checked with Coverity static analysis.
Deviations from the MISRA standard are listed below:

### Ignored by [Coverity Configuration](tools/coverity/misra.config)
| Deviation | Category | Justification |
| :-: | :-: | :-: |
| Directive 4.5 | Advisory | Allow names that MISRA considers ambiguous (such as LogInfo and LogError) |

> | Directive 4.8 | Advisory | Allow inclusion of unused types. Header files from coreMQTT, which are used by the agent files, may define types that are not used by the agent. |

Reviewer: It seems like the portion inside the commas just repeats information that's already there.
```suggestion
| Directive 4.8 | Advisory | Allow inclusion of unused types. Header files from coreMQTT may define types that are not used by the agent. |
```

Author: Sure

| Directive 4.9 | Advisory | Allow inclusion of function like macros. The `assert` macro is used throughout the library for parameter validation, and logging is done using function like macros. |

> | Rule 2.3 | Advisory | Allow unused types. Both the agent and coreMQTT headers define types intended for the application's use, which are not used by the agent files. |

Reviewer: I was a little confused by the wording because it seemed like the portion after the comma was describing `application's use`.
```suggestion
| Rule 2.3 | Advisory | Allow unused types. Both the agent and coreMQTT headers define types that are not directly used by the library but are still intended for the application's use. |
```

Author: Applied
